### PR TITLE
Fixed issue for new KDS with 32bit zynq

### PR DESCRIPTION
**Pull request by saifuddin-xilinx**
-- For 32bit Zynq new KDS is crashing.
-- On new kds mode we are creating a kernel thread of xrt_cu_polling_thread inside zocl_create_cu(zdev) function under a write lock write_lock(&zdev->attr_rwlock). Which is causing the issue.
-- Fixed this issue by moving zocl_create_cu(zdev) outside the write lock

### CODE REVIEW

File: src/runtime_src/core/edge/drm/zocl/zocl_xclbin.c [hunk 1]
```diff
@@ -935,6 +924,18 @@ zocl_xclbin_read_axlf(struct drm_zocl_dev *zdev, struct drm_zocl_axlf *axlf_obj,
...
+		ret = zocl_create_cu(zdev);
+		if (ret)
+			goto out0;
+		ret = zocl_kds_update(zdev);
+		if (ret)
+			goto out0;
+	}
+
 	zocl_clear_mem(zdev);
 	zocl_init_mem(zdev, zdev->topology);
```
Comment (mamin506) [inline src/runtime_src/core/edge/drm/zocl/zocl_xclbin.c:940]: This two function need write lock protection as well.

File: src/runtime_src/core/edge/drm/zocl/zocl_xclbin.c [hunk 2]
```diff
@@ -935,6 +924,18 @@ zocl_xclbin_read_axlf(struct drm_zocl_dev *zdev, struct drm_zocl_axlf *axlf_obj,
...
+
+	if (kds_mode == 1) {
+		subdev_destroy_cu(zdev);
+		ret = zocl_create_cu(zdev);
+		if (ret)
+			goto out0;
+		ret = zocl_kds_update(zdev);
+		if (ret)
+			goto out0;
+	}
```
Comment (mamin506) [inline src/runtime_src/core/edge/drm/zocl/zocl_xclbin.c:937]: Please leave a comment here to explain why this code block should be move out from the write lock region. Then others would not make same mistake and potentially can save a lot of time on debugging.

File: src/runtime_src/core/edge/drm/zocl/zocl_xclbin.c [hunk 3]
```diff
@@ -947,11 +948,15 @@ zocl_xclbin_read_axlf(struct drm_zocl_dev *zdev, struct drm_zocl_axlf *axlf_obj,
 	zdev->zdev_xclbin->zx_refcnt = 0;
 	zocl_xclbin_set_uuid(zdev, &axlf_head.m_header.uuid);
 
+	vfree(aie_res);
```
Comment (mamin506) [inline src/runtime_src/core/edge/drm/zocl/zocl_xclbin.c:951]: I dislike this duplicated code. Maybe we can use below code flow.

...
write_lock();
// Do everything that should be protected by the write lock

out0:
write_unlock();
if (!ret && kds_mode = 1) {
    // create cu sub-device
}
...

File: src/runtime_src/core/edge/drm/zocl/zocl_xclbin.c [hunk 4]
```diff
@@ -722,7 +722,6 @@ zocl_xclbin_read_axlf(struct drm_zocl_dev *zdev, struct drm_zocl_axlf *axlf_obj,
 		vfree(axlf);
 		return -EFAULT;
 	}
-
 	write_lock(&zdev->attr_rwlock);
```
Comment (mamin506) [inline src/runtime_src/core/edge/drm/zocl/zocl_xclbin.c:725]: Could you add a comment in before this lock to emphasize that "after this lock to unlock is an atomic context"?

**Review by mamin506 — CHANGES_REQUESTED** [review]

**Review by chvamshi-xilinx — APPROVED** [review]
Looks good. Can you please add a clear comment about why you had to make this change.
Like we are calling kthread_run from zocl_create_cu which is causing a kernel panic etc..

File: src/runtime_src/core/edge/drm/zocl/zocl_xclbin.c (hunk 2 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/edge/drm/zocl/zocl_xclbin.c]: Added a comments. Please check

File: src/runtime_src/core/edge/drm/zocl/zocl_xclbin.c (hunk 3 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/edge/drm/zocl/zocl_xclbin.c]: Modified accordingly

File: src/runtime_src/core/edge/drm/zocl/zocl_xclbin.c (hunk 4 above)
Comment (saifuddin-xilinx) [reply src/runtime_src/core/edge/drm/zocl/zocl_xclbin.c:727]: Added a comments

File: src/runtime_src/core/edge/drm/zocl/zocl_xclbin.c [hunk 5]
```diff
@@ -949,9 +940,22 @@ zocl_xclbin_read_axlf(struct drm_zocl_dev *zdev, struct drm_zocl_axlf *axlf_obj,
...
+			goto out1;
+		ret = zocl_kds_update(zdev);
+		if (ret)
+			goto out1;
+	}
+
+out1:
 	vfree(aie_res);
 	vfree(axlf);
-	DRM_INFO("%s %pUb ret: %d", __func__, zocl_xclbin_get_uuid(zdev), ret);
```
Comment (mamin506) [inline src/runtime_src/core/edge/drm/zocl/zocl_xclbin.c:954]: why remove this DRM_INFO?

**Review by mamin506 — CHANGES_REQUESTED** [review]

**Comment by mamin506** [comment]
> Looks good. Can you please add a clear comment about why you had to make this change.
> Like we are calling kthread_run from zocl_create_cu which is causing a kernel panic etc..

I don't think we should comment this directly reason in the code. The root cause is that current context could fall into sleep in a kernel API (kthread_run() in this case). In current code, we have to pay attention when add more code, must be atomic context friendly, in this write lock region.

**Comment by chvamshi-xilinx** [comment]
> > Looks good. Can you please add a clear comment about why you had to make this change.
> > Like we are calling kthread_run from zocl_create_cu which is causing a kernel panic etc..
> 
> I don't think we should comment this directly reason in the code. The root cause is that current context could fall into sleep in a kernel API (kthread_run() in this case). In current code, we have to pay attention when add more code, must be atomic context friendly, in this write lock region.

My point is same as "people has to pay attention when add more code". if we add a comment explicitly, people would think before adding a code. As this comes only for Zynq7000, We have high chances of breaking again.

**Comment by xilinxgitops** [comment]
retest this please

**Comment by mamin506** [comment]
> 
> My point is same as "people has to pay attention when add more code". if we add a comment explicitly, people would think before adding a code. As this comes only for Zynq7000, We have high chances of breaking again.

Agree 👍

**Review by mamin506 — APPROVED** [review]

**Pull request merged**